File: BucketSort.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

void bucketSort(vector<float> &A)
{
    vector<float> buckets[A.size()];
    // Putting each element into appropriate bucket
    for (int i = 0; i < A.size(); i++)
    {
        int bucketIndex = A[i] * A.size();
        buckets[bucketIndex].push_back(A[i]);
    }
    // Sorting elements in each bucket
    for (int i = 0; i < A.size(); i++)
        sort(buckets[i].begin(), buckets[i].end());
    int i = 0;
    // Concatenating all the sorted buckets
    for (int j = 0; j < A.size(); j++)
        for (int b = 0; b < buckets[j].size(); b++)
        {
            A[i] = buckets[j][b];
            i++;
        }
}

int main()
{
    vector<float> A = {0.32, 0.44, 0.31, 0.37, 0.17, 0.51, 0.18, 0.16, 0.17, 0.11};
    bucketSort(A);
    for (int i = 0; i < A.size(); i++)
        cout << A[i] << ' ';
    cout << endl;
    return 0;
}

File: BucketSort.java
//Bucket sort, or bin sort, is a sorting algorithm that works by distributing the elements of an array into a number of buckets.
//Each bucket is then sorted individually, either using a different sorting algorithm, or by recursively applying the bucket sorting algorithm.

//The Entire Bucket Sort Algorithm is written below in Java

import java.util.*;

public class BucketSort {
    public static void bucketSort(int[] arr) {
        int num_of_buckets = (int) Math.ceil(Math.sqrt(arr.length));
        int maxVal = Integer.MIN_VALUE;
        for (int element : arr) {
            if (element > maxVal) {
                maxVal = element;
            }
        }

        ArrayList<Integer>[] buckets = new ArrayList[num_of_buckets];

        for (int i = 0; i < buckets.length; i++) {
            buckets[i] = new ArrayList<Integer>();
        }

        for (int element : arr) {
            int bucketVal = (int) Math.ceil((float) element * num_of_buckets / maxVal);
            buckets[bucketVal - 1].add(element);
        }

        // Sort buckets
        for (int i = 0; i < num_of_buckets; i++) {
            System.out.println(buckets[i]);
            Collections.sort(buckets[i]);
        }
        int k = 0;
        for (int i = 0; i < num_of_buckets; i++) {
            for (int j = 0; j < buckets[i].size(); j++) {
                arr[k] = buckets[i].get(j);
                k++;
            }
        }
    }

    public static void printArrays(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

}

